src/main/java/datastructure/NodeGraph.java

```diff
@@ -0,0 +1,88 @@
+package datastructure;
+
+import java.util.ArrayList;
+
+/**
+ * Created by 101010.
+ */
+public class NodeGraph {
+    /**
+     * List of nodes.
+     */
+    private ArrayList<Node> nodes;
+
+    /**
+     * Database containing the segments of the nodes.
+     */
+    private SegmentDB segments;
+
+    /**
+     * Empty constructor for NodeGraph.
+     */
+    public NodeGraph() {
+        this.nodes = new ArrayList<Node>(0);
+        segments = new SegmentDB();
+    }
+
+    /**
+     * Constructor for NodeGraph
+     * @param nodes
+     *        The list of nodes for the graph.
+     * @param segments
+     *        The database containing the segments of the nodes.
+     */
+    public NodeGraph(final ArrayList<Node> nodes, final SegmentDB segments) {
+        this.nodes = nodes;
+        this.segments = segments;
+    }
+
+    /**
+     * Adds a node to the graph.
+     * @param id The id of the node.
+     * @param node The node that gets added.
+     * @param segment The segment of the node.
+     */
+    public void addNode(final int id, Node node, String segment) {
+        while (nodes.size() <= id) {
+            nodes.add(new Node());
+        }
+
+        int[] temp = nodes.get(id).getIncomingEdges();
+        nodes.set(id, node);
+        nodes.get(id).setIncomingEdges(temp);
```

Comment:
We can get incoming edges of a node before we find that node, thus instead we add the incoming edges to a temporary node and then when we find that node we set its incoming edges to the incoming edges we have found for it so far.